api/python/t4/formats.py

```diff
@@ -0,0 +1,505 @@
+        """Match a format (or formats) by extension."""
+        if not ext:
+            return None if single else []
+        ext = ext.lower().strip('. ')
+        matching_formats = cls.formats_by_ext[ext]
+        if single:
+            return matching_formats[0] if matching_formats else None
+        return matching_formats[:]
+
+    @classmethod
+    def for_obj(cls, obj, single=True):
+        """Match a format (or formats) by a (potentially) serializable object"""
+        formats = reversed(cls.registered_formats.values())
+        if single:
+            for format in formats:
+                if format.handles_obj(obj):
+                    return format
+            return
+        return [format for format in formats if format.handles_obj(obj)]
+
+    @classmethod
+    def _get_name_from_meta(cls, meta):
+        name = meta.get('format', {}).get('name')
+        # 'target': compat with older pkg structure -- can probably be removed soon.
+        if not name:
+            name = meta.get('target')
+        return name
+
+    @classmethod
+    def for_meta(cls, meta):
+        """Unambiguously match a specific format by the given metadata"""
+        # As a point of order, this must return a singular format.
+        name = cls._get_name_from_meta(meta)
+        fmt = cls.match(name)
+        return fmt
+
+
+class BaseFormat(ABC):
+    """Base class for binary format handlers
+    """
+    def __init__(self, name=None, handled_extensions=None, handled_types=None):
+        """Common initialization for BaseFormat subclasses
+        
+        Subclasses implement the `serialize()` and `deserialize()` methods,
+        which are passed the object/bytes to handle, as well as metadata and
+        runtime kwargs.
+        
+        Subclasses *may* implement custom `handles_ext`, `handles_obj` methods
+        if there is a scenario which requires it (such as lazy load of a large
+        module).
+        
+        Args:
+            name(str): Name of new format.  Use existing name if your
+                format is compatible with existing formats, if practicable.  
+                I.e., two different CSV format handlers should both use 'csv'.
+
+            handled_extensions(list(str)): a list of filename extensions
+                that can be deserialized by this format
+
+            handled_types: a list of types that can be serialized to
+                (and deserialized from) by this format
+        """
+        self.name = name if name else getattr(self, 'name', None)
+        if not self.name:
+            raise TypeError("No `name` attribute has been defined for {!r}".format(type(self).__name__))
+
+        # add user extensions if given
+        handled_extensions = [] if handled_extensions is None else handled_extensions
+        self.handled_extensions = getattr(self, 'handled_extensions', [])
+        self.handled_extensions.extend(ext for ext in handled_extensions if ext not in self.handled_extensions)
```

Comment:
This is O(N²)... but really, I'd rather get rid of the `getattr` above.

Reply:
Sets are overkill, but sure.

Reply:
done